### PR TITLE
Improve job system throughput by 1.5x

### PR DESCRIPTION
**Pull request by jeremyong-az**
The mechanism used to notify threads that work is available is driven
by AZStd::binary_semaphore which on Windows is backed by a Win32 event.
Win32 events are a kernel space object that can be shared across
processes, and are generally expensive. In contrast,
AZStd::condition_variable is backed by a Win32 CONDITION_VARIABLE. While
a mutex is needed to wait on the condition variable, the mutex is
generally uncontended. More importantly, the CONDITION_VARIABLE is a
userspace object (instead of a kernel space object), and the uncontended
mutex will often be released immediately without entering the kernel.

The result of a simple microbenchmark demonstrates that enqueueing a
single job based on the old mechanism takes on the order of 400 ns. However,
changing the notification mechanism to use AZStd::condition_variable
instead drops this time to ~285 ns.

More complete benchmarking results are provided below, based on the
existing benchmarking tests. The columns correspond to Total Time(ns),
CPU(ns), and iteration count per Google's benchmark library. The second
column is the important one to compare (lower is better).

## OLD RESULTS (binary_semaphore)

JobBenchmarkFixture/RunSmallNumberOfLightWeightJobsWithDefaultPriority 33179 ns        32959 ns        21333
JobBenchmarkFixture/RunMediumNumberOfLightWeightJobsWithDefaultPriority 3362169 ns      3370098 ns          204
JobBenchmarkFixture/RunLargeNumberOfLightWeightJobsWithDefaultPriority 55392209 ns     55397727 ns           11
JobBenchmarkFixture/RunSmallNumberOfMediumWeightJobsWithDefaultPriority 32737 ns        32226 ns        20364
JobBenchmarkFixture/RunMediumNumberOfMediumWeightJobsWithDefaultPriority 3321626 ns      3293505 ns          204
JobBenchmarkFixture/RunLargeNumberOfMediumWeightJobsWithDefaultPriority 52722218 ns     52556818 ns           11
JobBenchmarkFixture/RunSmallNumberOfHeavyWeightJobsWithDefaultPriority 1798466 ns      1765971 ns          407
JobBenchmarkFixture/RunMediumNumberOfHeavyWeightJobsWithDefaultPriority 117518267 ns    117187500 ns            6
JobBenchmarkFixture/RunLargeNumberOfHeavyWeightJobsWithDefaultPriority 1864236000 ns   1828125000 ns            1
JobBenchmarkFixture/RunSmallNumberOfRandomWeightJobsWithDefaultPriority 916962 ns       899431 ns          747
JobBenchmarkFixture/RunMediumNumberOfRandomWeightJobsWithDefaultPriority 57428870 ns     54687500 ns           10
JobBenchmarkFixture/RunLargeNumberOfRandomWeightJobsWithDefaultPriority 925064600 ns    890625000 ns            1
JobBenchmarkFixture/RunSmallNumberOfLightWeightJobsWithRandomPriorities 33970 ns        34528 ns        20364
JobBenchmarkFixture/RunMediumNumberOfLightWeightJobsWithRandomPriorities 3526404 ns      3523284 ns          204
JobBenchmarkFixture/RunLargeNumberOfLightWeightJobsWithRandomPriorities 54560909 ns     53977273 ns           11
JobBenchmarkFixture/RunSmallNumberOfMediumWeightJobsWithRandomPriorities 34157 ns        34424 ns        21333
JobBenchmarkFixture/RunMediumNumberOfMediumWeightJobsWithRandomPriorities 3398680 ns      3446691 ns          204
JobBenchmarkFixture/RunLargeNumberOfMediumWeightJobsWithRandomPriorities 58225940 ns     57812500 ns           10
JobBenchmarkFixture/RunSmallNumberOfHeavyWeightJobsWithRandomPriorities 1903216 ns      1881143 ns          407
JobBenchmarkFixture/RunMediumNumberOfHeavyWeightJobsWithRandomPriorities 118144683 ns    114583333 ns            6
JobBenchmarkFixture/RunLargeNumberOfHeavyWeightJobsWithRandomPriorities 1869486300 ns   1671875000 ns            1
JobBenchmarkFixture/RunSmallNumberOfRandomWeightJobsWithRandomPriorities 928852 ns       941265 ns          747
JobBenchmarkFixture/RunMediumNumberOfRandomWeightJobsWithRandomPriorities 58163618 ns     53977273 ns           11
JobBenchmarkFixture/RunLargeNumberOfRandomWeightJobsWithRandomPriorities 932691200 ns    843750000 ns            1

## NEW RESULTS (condition_variable)

JobBenchmarkFixture/RunSmallNumberOfLightWeightJobsWithDefaultPriority 25129 ns        24902 ns        26353
JobBenchmarkFixture/RunMediumNumberOfLightWeightJobsWithDefaultPriority 2234018 ns      2247074 ns          299
JobBenchmarkFixture/RunLargeNumberOfLightWeightJobsWithDefaultPriority 33588352 ns     34226190 ns           21
JobBenchmarkFixture/RunSmallNumberOfMediumWeightJobsWithDefaultPriority 27176 ns        27274 ns        26353
JobBenchmarkFixture/RunMediumNumberOfMediumWeightJobsWithDefaultPriority 2179414 ns      2176339 ns          280
JobBenchmarkFixture/RunLargeNumberOfMediumWeightJobsWithDefaultPriority 33352230 ns     33593750 ns           20
JobBenchmarkFixture/RunSmallNumberOfHeavyWeightJobsWithDefaultPriority 1971620 ns      1947464 ns          345
JobBenchmarkFixture/RunMediumNumberOfHeavyWeightJobsWithDefaultPriority 119505200 ns    113839286 ns            7
JobBenchmarkFixture/RunLargeNumberOfHeavyWeightJobsWithDefaultPriority 1913765800 ns   1828125000 ns            1
JobBenchmarkFixture/RunSmallNumberOfRandomWeightJobsWithDefaultPriority 978376 ns       976562 ns          640
JobBenchmarkFixture/RunMediumNumberOfRandomWeightJobsWithDefaultPriority 58535218 ns     46875000 ns           11
JobBenchmarkFixture/RunLargeNumberOfRandomWeightJobsWithDefaultPriority 944719700 ns    703125000 ns            1
JobBenchmarkFixture/RunSmallNumberOfLightWeightJobsWithRandomPriorities 25626 ns        25844 ns        23579
JobBenchmarkFixture/RunMediumNumberOfLightWeightJobsWithRandomPriorities 2156639 ns      2142559 ns          299
JobBenchmarkFixture/RunLargeNumberOfLightWeightJobsWithRandomPriorities 34256847 ns     33717105 ns           19
JobBenchmarkFixture/RunSmallNumberOfMediumWeightJobsWithRandomPriorities 27427 ns        27169 ns        23579
JobBenchmarkFixture/RunMediumNumberOfMediumWeightJobsWithRandomPriorities 2210968 ns      2247074 ns          299
JobBenchmarkFixture/RunLargeNumberOfMediumWeightJobsWithRandomPriorities 33389135 ns     33593750 ns           20
JobBenchmarkFixture/RunSmallNumberOfHeavyWeightJobsWithRandomPriorities 2001554 ns      1992754 ns          345
JobBenchmarkFixture/RunMediumNumberOfHeavyWeightJobsWithRandomPriorities 121427817 ns    109375000 ns            6
JobBenchmarkFixture/RunLargeNumberOfHeavyWeightJobsWithRandomPriorities 1931018800 ns   1562500000 ns            1
JobBenchmarkFixture/RunSmallNumberOfRandomWeightJobsWithRandomPriorities 999190 ns      1004016 ns          747
JobBenchmarkFixture/RunMediumNumberOfRandomWeightJobsWithRandomPriorities 58714218 ns     49715909 ns           11
JobBenchmarkFixture/RunLargeNumberOfRandomWeightJobsWithRandomPriorities 947552300 ns    671875000 ns            1

Signed-off-by: Jeremy Ong <jcong@amazon.com>

### CODE REVIEW

File: Code/Framework/AzCore/Tests/Jobs.cpp [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include <AzCore/std/containers/array.h>
 #include <AzCore/std/containers/fixed_list.h>
 #include <AzCore/std/containers/unordered_set.h>
+#include <AzCore/std/parallel/binary_semaphore.h>
```
Comment (jeremyong-az) [inline Code/Framework/AzCore/Tests/Jobs.cpp:26]: Note, this header is needed for some of the tests, but no longer included in `JobManagerWorkStealing.h`

File: Code/Framework/AzCore/Tests/Jobs.cpp [hunk 2]
```diff
@@ -1232,7 +1234,7 @@ namespace UnitTest
             }
 #endif // AZ_COMPARE_TO_PPL
 
-            AZ_Printf("UnitTest", "\n\nJob overhead test. Serial %lld (%lld) Parallel %lld (%lld) PPL %lld (%lld) Total: %lld\n\n", nonParallelMS, nonParallelProcessMS, parallelForMS, parallelForProcessMS, parallelForPPLMS, parallelForProcessPPLMS, s_totalJobsTime);
+            printf("\n\nJob overhead test. Serial %lld (%lld) Parallel %lld (%lld) PPL %lld (%lld) Total: %lld\n\n", nonParallelMS, nonParallelProcessMS, parallelForMS, parallelForProcessMS, parallelForPPLMS, parallelForProcessPPLMS, s_totalJobsTime);
```
Comment (rgba16f) [inline Code/Framework/AzCore/Tests/Jobs.cpp:1237]: This looks like a left over debugging change

**Review by rgba16f — APPROVED** [review]

File: Code/Framework/AzCore/Tests/Jobs.cpp [hunk 3]
```diff
@@ -220,6 +221,7 @@ namespace UnitTest
 
             //test user jobs
             {
+
```
Comment (amznestebanpapp) [inline Code/Framework/AzCore/Tests/Jobs.cpp:224]: delete?

File: Code/Framework/AzCore/Tests/Jobs.cpp (hunk 2 above)
Comment (amznestebanpapp) [inline Code/Framework/AzCore/Tests/Jobs.cpp:1237]: seems this was for debugging? should be reverted

**Comment by jeremyong-az** [comment]
NOTE: don't check in until a followup commit addressing compile errors due to downstream headers/source files relying on `binary_semaphore.h` (which was removed)

File: Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp [hunk 4]
```diff
@@ -98,8 +98,7 @@ JobManagerWorkStealing::~JobManagerWorkStealing()
 
         for (ThreadInfo* thread : m_workerThreads)
         {
-            thread->m_waitEvent.release();
-            thread->m_thread.join();
```
Comment (amznestebanpapp) [inline Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp:102]: I think this join was there to finish executing those threads during shutdown.
Note the `m_quitRequested` being set here and the `if` on line 376
I think the original method is also unsafe... it should signal and wait for the thread to finish

File: Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp [hunk 5]
```diff
@@ -370,7 +369,8 @@ void JobManagerWorkStealing::ProcessJobsInternal(ThreadInfo* info, Job* suspende
                 if (shouldSleep)
                 {
                     //no available work, so go to sleep (or we have already been signaled by another thread and will acquire the semaphore but not actually sleep)
-                    info->m_waitEvent.acquire();
+                    AZStd::unique_lock<AZStd::mutex> lock{ info->m_waitMutex };
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp:372]: FYI Class Template Argument Deduction gets rid of the need to specify `<AZStd::mutex>`

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp (hunk 5 above)
Comment (jeremyong-az) [reply Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp:372]: Thanks will change I keep forgetting we're on C++17 (worked on a C++14 codebase most recently). CTAD is indeed great

File: Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp (hunk 4 above)
Comment (jeremyong-az) [reply Code/Framework/AzCore/AzCore/Jobs/Internal/JobManagerWorkStealing.cpp:102]: This was a typo, cleaning up the commit and everything now thanks

**Review by amznestebanpapp — APPROVED** [review]

**Review by lawsonamzn — APPROVED** [review]

**Review by kberg-amzn — APPROVED** [review]
Looks great!

Since we control AZStd::binary_semaphore, and I very much doubt any usage requires or depends on IPC, maybe we should replace the guts of AZStd::binary_semaphore with a condition variable like you've done here?

**Comment by jeremyong-az** [comment]
> Looks great!
> 
> Since we control AZStd::binary_semaphore, and I very much doubt any usage requires or depends on IPC, maybe we should replace the guts of AZStd::binary_semaphore with a condition variable like you've done here?

Yea I agree didn't want to do such an invasive change here just yet (and it would change the size of the binary_semaphore) but we should definitely investigate that option

**Review by AMZN-koppersr — APPROVED** [review]

**Comment by jeremyong** [comment]
So, the build failed due to a timeout in the linux branch. But this code has literally no bearing on Linux (all of it is behind a windows pound def). It may just be unlucky? I'll kick it off again.

**Comment by jeremyong-az** [comment]
Ah, I messed up, I should have looked at the implementation of our condition_variable more carefully. It doesn't spin to check for spurious wakeups on behalf of the user.

**Comment by jeremyong-az** [comment]
Closing! In favor of a different PR I will put up in the near future :)

So, interestingly, my intuition about Win32 Events is slightly out-of-date. Once I account for spurious wakeups (user error, the condition variable class I used in a previous codebase did this internally), the speedup is present but small enough that it could be attributable to noise. As an experiment, I reimplemented `binary_semaphore` to the following:

```c++
    class binary_semaphore
    {
    public:
        binary_semaphore(bool initialState = false)
            : m_state{ initialState }
        {
        }

        binary_semaphore(const char* name, bool initialState = false)
            : m_state{ initialState }
        {
            (void)name;
        }

        binary_semaphore(const binary_semaphore&) = delete;
        binary_semaphore& operator=(const binary_semaphore&) = delete;

        void acquire()
        {
            while (true)
            {
                bool result = m_state.load(memory_order_relaxed);
                while (!m_state)
                {
                    WaitOnAddress(&m_state, &result, sizeof(m_state), AZ_INFINITE);
                    result = true;
                }

                if (m_state.compare_exchange_strong(result, false))
                {
                    return;
                }
            }
        }

        void release()
        {
            m_state.store(true, memory_order_release);
            WakeByAddressSingle(&m_state);
        }

private:
        atomic<bool> m_state;
};
```

Even with this lean implementation, the improvement I'm seeing is nominal (if these benchmarks are to be believed). The issue with our `condition_variable` implementation which I believe we should investigate in the future is the use of a normal mutex instead of an SRW mutex (slim reader writer) [link](https://docs.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks). Interestingly, I decided to verify this implementation against the C++20 atomic wait implementation in the STL, and surprisingly, they spin! For reference: https://github.com/microsoft/STL/blob/9a9820df1a1d3fa84100e3169ff37fdd4fa41759/stl/inc/atomic#L580 

In other words, I apparently have no idea what I'm talking about, at least with more recent versions of the kernel so I'm going to look forward to the more important optimization of relaxing prioritization granularity instead.
